syntax = "proto3";

package platform.v1;

import "google/protobuf/timestamp.proto";

// User ma'lumotlari uchun message
message User {
    int64 id = 1;
    string name = 2;
    string email = 3;
    int32 age = 4;
    bool is_active = 5;
    google.protobuf.Timestamp created_at = 6;
}

// Product ma'lumotlari uchun message
message Product {
    int64 id = 1;
    string name = 2;
    string description = 3;
    double price = 4;
    string category = 5;
    int32 quantity = 6;
    google.protobuf.Timestamp created_at = 7;
}

// Order ma'lumotlari uchun message
message Order {
    int64 id = 1;
    int64 user_id = 2;
    repeated OrderItem items = 3;
    double total_amount = 4;
    OrderStatus status = 5;
    google.protobuf.Timestamp created_at = 6;
}

// Order item ma'lumotlari
message OrderItem {
    int64 product_id = 1;
    string product_name = 2;
    int32 quantity = 3;
    double unit_price = 4;
    double total_price = 5;
}

// Order statuslari
enum OrderStatus {
    ORDER_STATUS_PENDING = 0;
    ORDER_STATUS_CONFIRMED = 1;
    ORDER_STATUS_SHIPPED = 2;
    ORDER_STATUS_DELIVERED = 3;
    ORDER_STATUS_CANCELLED = 4;
}

// Webhook uchun asosiy data message
message WebhookData {
    string event_type = 1;  // "user_created", "order_placed", "product_updated"
    string event_id = 2;
    google.protobuf.Timestamp timestamp = 3;
    
    // oneof - faqat bitta field bo'ladi
    oneof data {
        User user_data = 4;
        Product product_data = 5;
        Order order_data = 6;
    }
}

// Response message webhook uchun
message WebhookResponse {
    bool success = 1;
    string message = 2;
    google.protobuf.Timestamp processed_at = 3;
}